### PR TITLE
Gsdc2021 debugging

### PR DESCRIPTION
**Pull request by plutonheaven**
This branch applies `prx` to RNX files from the [Google Smartphone Decimeter Challenge 2021](https://www.kaggle.com/c/google-smartphone-decimeter-challenge).

This is a good test to check that `prx` can work with files from other sources than the high quality IGS network. For example, tthe GSDC RNX files have missing header information (leap second, time of last obs, approximate ecef position...).

PRX was able to process the 121 files of the GSDC2021 dataset in 4h10, (27 min when using parallel loops on a 36-core computer).

### CODE REVIEW

File: src/prx/helpers.py [hunk 1]
```diff
@@ -362,10 +368,16 @@ def cached_load(rinex_file: Path, file_hash: str):
 
 def get_gpst_utc_leap_seconds_from_rinex_header(rinex_file: Path):
     header = georinex.rinexheader(rinex_file)
-    assert "LEAP SECONDS" in header, "LEAP SECONDS not found in RINEX header"
-    ls_before = header["LEAP SECONDS"][0:6].strip()
+
+    if "LEAP SECONDS" not in header:
+        # Leap second not in RNX NAV file, get it from astropy
+        return compute_gps_leap_seconds(
```
[the remark this answers is not in the document]
Comment (plutonheaven) [reply src/prx/helpers.py:374]: If we assume that the NAV files are downloaded from IGS servers, then, I would say that the name can be trusted. If we let the user provide the NAV file (e.g. in a future command line option), then, not sure about it.

The TIME OF FIRST OBS is more reliable, but is a field from the OBS file.

**Comment by plutonheaven** [comment]
The script has been able to process all 121 RNX OBS file from the dataset!! 🎉

It is not that long, 2 to 4 minutes per file, each of it is several tens of minutes at 1 Hz.

I'll do some additional checks and improvements before setting this PR as ready.

**Comment by plutonheaven** [comment]
PRX was able to process the 121 files of the GSDC2021 dataset in 4h10, (27 min when using parallel loops on a 36-core computer).

File: src/prx/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (plutonheaven) [reply src/prx/helpers.py:374]: Thinking about your comment, here are my thoughts:
- currently, PRX is called on a RINEX OBS file
- based on the `TIME OF FIRST/LAST OBS` in the header, the required RINEX NAV files are downloaded from IGS servers
- files from IGS should meet a high quality standard, in terms of how they are generated, so I would believe that we can trust the name of the file.

Later, if we want to let the user define its own NAV file, then we could run in some issues (we could still probably rename the NAV file with `gfzrnx` using the RNX3 convention). I'll open an issue to remember this.

File: .gitignore [hunk 2]
```diff
@@ -2,14 +2,17 @@
...
 *.csv
+*.prof
 *.json*
 */tmp_*
-/datasets/TLSE_2022001/BRDC00IGS_R_20220010000_01D_MN.rnx
-/datasets/TLSE_2022001/BRDC00IGS_R_20220010000_01D_GN.rnx
-/datasets/TLSE_2022001/BRD400IGS_S_20220010000_01D_MN.rnx
-/datasets/TLSE_2022001/TLSE00FRA_R_20220010000_01D_30S_MO.rnx
-/datasets/TLSE_2022001/TLSE00FRA_R_20220010000_01D_30S_GO.rnx
+/src/prx/test/datasets/TLSE_2022001/BRDC00IGS_R_20220010000_01D_MN.rnx
```
Comment (jtec) [inline .gitignore:8]: can we use a regex here, something like `/src/prx/test/datasets/*.rnx?` I think we ususally track only `.CRX.gz` files, right?

File: poetry.lock
Comment (jtec) [inline poetry.lock:1]: Looks like there migth be a few unrelated changes in the lock file - could you try to check out the `poetry.lock` on `main` and run `poetry lock --no-update`? That way poetry should only add new stuff in `pyproject.toml` and not update other packages as well. The idea is to keep changes in each PR as small and focused as possible.

File: .gitignore (hunk 2 above)
Comment (plutonheaven) [reply .gitignore:8]: I'll ignore all `*.rnx` and `*.crx` files.

File: poetry.lock
Comment (plutonheaven) [reply poetry.lock:1]: Thanks for the advice. I wasn't sure about all the update I had in `poetry.lock` after adding `astropy` to the dependencies. Today, I learned... 🙏

File: src/dataset_processing_script/main_gsdc2021.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    # Parallel(n_jobs=1,verbose=100)(
+    #     delayed(prx.process)(
+    #         observation_file_path=file,
+    #         output_format="csv",
+    #     )
+    #     for file in local_gz_list
+    # )
+
+
+if __name__ == "__main__":
```
Comment (jtec) [inline src/dataset_processing_script/main_gsdc2021.py:62]: Can we make the path a command-line argument?

File: src/dataset_processing_script/main_gsdc2021.py
Comment (jtec) [inline src/dataset_processing_script/main_gsdc2021.py:1]: We could rename the folder `examples` or `demos` as it demonstrates one use case of `prx`.

File: src/prx/rinex_nav/evaluate.py [hunk 4]
```diff
@@ -639,6 +639,17 @@ def compute_parallel(rinex_nav_file_path, per_signal_query):
 
 
 def compute(rinex_nav_file_path, per_signal_query):
+    # per_signal_query is a pd.DataFrame with the following columns
+    #   - time_of_reception_in_receiver_time
+    #   - satellite
+    #   - observation_value
+    #   - observation_type
+    #   - time_of_emission_in_satellite_time_integer_second_aligned_to_receiver_time
```
Comment (jtec) [inline src/prx/rinex_nav/evaluate.py:647]: missing item?

File: src/prx/test/test_helpers.py [hunk 5]
```diff
@@ -297,3 +297,47 @@ def test_row_wise_dot_product():
...
+    # expected values come from https://gnsscalc.com/
+    week_expected = [54, 54, 1025, 2048, np.nan]
+    seconds_of_week_expected = [280800, 281400, 317400, 302400, np.nan]
+
+    assert week_expected == week_computed
+    assert seconds_of_week_expected == seconds_of_week_expected
+
+
+def test_compute_gps_leap_seconds():
+    # expected GPS lepqp second come from https://gnsscalc.com/
```
Comment (jtec) [inline src/prx/test/test_helpers.py:330]: typo

File: src/prx/test/test_helpers.py [hunk 6]
```diff
@@ -297,3 +297,47 @@ def test_row_wise_dot_product():
...
     assert (row_wise_dot == np.array([[10 + 40], [120 + 200], [350 + 480]])).all()
+
+
+def test_timedelta_2_weeks_and_seconds():
+    tested_timestamps = [
+        pd.Timestamp("1981-01-21T06:00:00"),
+        pd.Timestamp("1981-01-21T06:10:00"),
+        pd.Timestamp("1999-09-01T16:10:00"),  # after first week number rollover
+        pd.Timestamp("2019-04-10T12:00:00"),  # after 2nd week number rollover
+        pd.Timestamp(""),  # yiels a NaT
```
Comment (jtec) [inline src/prx/test/test_helpers.py:308]: typo

File: src/prx/user.py [hunk 7]
```diff
@@ -106,3 +108,135 @@ def spp_pt_lsq(df, dx_convergence_l2=1e-6, max_iterations=10):
             n_iterations <= max_iterations
         ), "LSQ did not converge in allowed number of iterations"
     return x_linearization
+
+
+def first_position_from_georinex(filepath_obs, filepath_nav):
```
Comment (jtec) [inline src/prx/user.py:113]: Suggest something like `bootstrap_coarse_receiver_position`, that way the function name is independent of the RINEX parser we use.

**Comment by jtec** [comment]
Great stuff! Left a few minor comments, but looks good to me!

**Review by jtec — APPROVED** [review]

File: src/dataset_processing_script/main_gsdc2021.py (hunk 3 above)
Comment (plutonheaven) [reply src/dataset_processing_script/main_gsdc2021.py]: Done in e6d3274c1d88b4de0328b6e9b2fcd0029979e7a2
